File: Leetcode/_5_Longest_Palindromic_Substring.java
// link - https://leetcode.com/problems/longest-palindromic-substring/

public class _5_Longest_Palindromic_Substring {
    public static void main(String[] args) {
        String s = "babad"; // ans = bab OR aba
        String s2 = "cbbd"; // ans = bb

        System.out.println(longestPalindromeManachers(s));
        System.out.println(longestPalindromeManachers(s2));
    }

    // Manacher's Algorithm - Time = O(n) - Space = O(n)

    // (The code for manachers is hella unoptimised, retry this and make it more optimized when you revisit this one)

    // put a unique string like "#" between the string so that we are technically
    // calculating
    // odd palindromes
    // iterate over each char
    // keep track of l and r denoting the left and right bounds of the current
    // substring
    // the idea is that the length of the substring to the left of the current
    // parent substring will be the same or greater for the sub palindromes to the
    // right
    // check if we have already calculated and assign the max possible value of the
    // palindrome
    // to the one at the right which will be min(r - l, (mid - (r - mid)))
    // thats becuase anything that is after r is unexplored sso we dont know if it
    // will make a
    // palindrome or not
    // so after that, we will manually check whether we can make a bigger palindrome
    // or not
    // and in each iteration we check whether current one is the biggest palindrome,
    // if yes then return that substring at the end
    // and thats it!

    // 0 1 2 3 4 5 6
    // m i r
    // m - (i - m) = 3 - (5 - 3) = 3 - 2 = 1

    // # 0 # 1 # 2 # 3 # 4 # 5 # 6 #

    public static String longestPalindromeManachers(String s) {
        if (s.length() <= 1) {
            return s;
        }

        // taking the separating character as "#" since the input is alphanumeric
        String str = s.replaceAll("", "#");

        int[] count = new int[str.length()];

        int maxSize = 0;
        int maxIdx = 0;

        // m is the middle of the current largest palindrome
        // r is the rightmost element of the current largest palindrome
        // we dont need an l variable because we can just calculate it
        // i is the current element we are checking
        int r = 0;
        int m = 0;
        for (int i = 0; i < str.length(); i++) {
            if (i < r) {
                count[i] = Math.min(r - i, count[2 * m - i]);
            }

            int a = i - count[i];
            int b = count[i] + i;
            while (a >= 0 && b < count.length && (str.charAt(a) == str.charAt(b))) {
                count[i]++;
                a--;
                b++;
            }

            if (b > r) {
                m = i;
                r = b - 1;
            }

            if (count[i] > maxSize) {
                maxSize = count[i];
                maxIdx = i;
            }
        }

        return str.substring(maxIdx - maxSize + 1, maxIdx + maxSize).replace("#", "");
    }

    // Expand Around Center approach - Time = O(n^2) - Space = O(1)
    public String longestPalindromeEAC(String s) {
        if (s.length() <= 1)
            return s;

        int[] max = { 0, 1 };
        for (int i = 0; i < s.length(); i++) {
            int[] even = isPalindromeEAC(s, i, i);
            int[] odd = isPalindromeEAC(s, i, i + 1);

            if (even[1] - even[0] > max[1] - max[0]) {
                max = even;
            }
            if (odd[1] - odd[0] > max[1] - max[0]) {
                max = odd;
            }
        }

        return s.substring(max[0], max[1]);
    }

    public int[] isPalindromeEAC(String s, int l, int r) {
        while (l >= 0 && r < s.length() && (s.charAt(l) == s.charAt(r))) {
            l--;
            r++;
        }
        return new int[] { l + 1, r };
    }

    // Naive Solution - Time = O(n^3) - Space = O(1)
    public String longestPalindrome(String s) {
        int max = 0;
        int ansStart = 0, ansEnd = 0;
        for (int i = 0; i < s.length(); i++) {
            for (int j = i + 1; j < s.length(); j++) {
                if (isPalindrome(s, i, j)) {
                    if (j - i > max) {
                        max = j - i;
                        ansStart = i;
                        ansEnd = j;
                    }
                }
            }
        }
        return s.substring(ansStart, ansEnd + 1);
    }

    public boolean isPalindrome(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l) != s.charAt(r))
                return false;
            l++;
            r--;
        }
        return true;
    }
}

File: Leetcode/python/_5_Longest_Palindromic_Substring.py
class Solution:
    # Manacher's Algorithm - Time = O(n) - Space = O(n)
    # This is an optimized version of expand-around-center that avoids redundant comparisons
    # by reusing palindrome information through symmetry.
    #
    # We first transform the string by inserting separators (#) so that all palindromes
    # have odd length, allowing uniform handling of even and odd palindromes.
    #
    # While iterating, we maintain the center (c) and right boundary (r) of the rightmost
    # palindrome found so far. If the current index lies within this boundary, we can
    # initialize its palindrome radius using the mirrored index across c, clamped by r.
    #
    # We then expand only beyond the known boundary, ensuring each character is compared
    # at most once overall, which guarantees linear time complexity.
    def longestPalindrome(self, s: str) -> str:
        # Put # between each letter and the boundaries so that the total length is always consistent
        pal = "^#" + "#".join(s) + "#$"
        n = len(pal)
        # Length will track the span of the palindrome starting from each letter
        length = [0] * n

        # c will track the center of the biggest palindrome we've seen till now, r will track its right boundary
        c = r = 0
        for i in range(1, n - 1):
            # First find the index of element that mirrors current element i
            mirror = 2 * c - i

            # If the current index we're checking is within the span of the current biggest palindrome
            # found, then we can depend on the mirror property of a palindrome, and this ensures that
            # either the minimum of the remaining part from i to r, or the length of palindrome
            # originating from current elements mirror, will be the same
            if i < r:
                length[i] = min(r - i, length[mirror])

            # Start expanding from the next unknown index
            while pal[i - length[i] - 1] == pal[i + length[i] + 1]:
                length[i] += 1

            # If we find a bigger r boundary then update it
            if i + length[i] > r:
                c = i
                r = i + length[i]

        # In the end we'll find the biggest palindrome center, convert its dimensions for the original
        # string, and return that substring
        max_len = max(length)
        center = length.index(max_len)
        start = (center - max_len) // 2
        return s[start : start + max_len]

    # Expand Around Center - Time = O(n^2) - Space = O(1)
    # For each letter expand around it until you find a mismatching corresponding character and track the max
    # While expanding the palindrome could start at a single character ("bob") or two characters ("lool")
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        res = ""

        for i in range(n):
            # Check for one char center palindromes
            l, r = i, i
            while l >= 0 and r < n and s[l] == s[r]:
                if len(res) < r - l + 1:
                    res = s[l : r + 1]
                l -= 1
                r += 1

            # Check for two char center palindromes
            l, r = i, i + 1
            while l >= 0 and r < n and s[l] == s[r]:
                if len(res) < r - l + 1:
                    res = s[l : r + 1]
                l -= 1
                r += 1

        return res
